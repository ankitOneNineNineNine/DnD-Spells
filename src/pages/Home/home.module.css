.spellContainer {
  display: "flex";
  flex-wrap: "wrap";
  height: 75vh;
  overflow-y: scroll;
  margin-top: 10px;
}

.spellContainer {
  -ms-overflow-style: none;
}

.spellContainer > a {
  color: white;
  text-decoration: none;
}
.favoriteButton {
  background: #6c4470;
  padding: 10px;
  border: 5px;
  border-radius: 5px;
  color: white;
  cursor: pointer;
}
