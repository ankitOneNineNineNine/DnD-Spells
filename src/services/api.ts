/**
 * Apis List
 * @export api
 */
export const api = {
  spells: {
    fetch: "https://www.dnd5eapi.co/api/spells",
    fetchById: "https://www.dnd5eapi.co/api/spells/{id}"
  }
};
