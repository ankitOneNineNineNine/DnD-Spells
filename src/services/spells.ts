import axios, { AxiosError } from "axios";
import { useQuery } from "react-query";
import { api } from "./api";
import { toastFail } from "./toast";

export interface ISpell {
  index: string;
  name: string;
  url: string;
}
interface ISchool {
  index: string;
  name: string;
  url: string;
}

type IClass = ISpell;
type ISubClass = ISpell;
type IDamageType = ISpell;
export const Componentsenum: { [key: string]: any } = {
  V: "verbal",
  S: "somatic",
  M: "material"
};
export interface ISpellInfo {
  casting_time?: string;
  classes?: IClass[];
  components?: string[];
  concentration?: boolean;
  desc?: string[];
  duration?: string;
  damage?: {
    damage_at_character_level?: { [key: number]: string };
    damage_type?: IDamageType;
  };
  area_of_effect?: {
    size?: number;
    type?: string;
  };
  attack_type?: string;
  higher_level?: string[];
  index?: string;
  level?: 2;
  material?: string;
  name?: string;
  range?: string;
  ritual?: boolean;
  school?: ISchool;
  subclasses?: ISubClass[];
  url?: string;
  _id?: string;
}

export interface IFetchSpellsResponse {
  count: number;
  results: ISpell[];
}

/**
 * Axios Get Call to fetch repos based on search query
 *
 * @returns AxiosResponse<IFetchSpellsResponse>
 */
const fetchSpells = () => () => {
  return axios.get<IFetchSpellsResponse>(api.spells.fetch);
};

/**
 * Axios Get Call to get Spell by Index
 * @param {string} id
 * @returns AxiosResponse<ISpell>
 */
const getSpellByIndex = (id: string) => () => {
  return axios.get<ISpellInfo>(api.spells.fetchById.replace("{id}", id));
};

/**
 * Use Query to Fetch All Spells
 * @export useGetAllSpells
 * @returns UseQueryResult
 */
export const useGetAllSpells = () => {
  return useQuery(api.spells.fetch, fetchSpells(), {
    select: (response) => response.data,
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toastFail(error?.response?.data?.message || "Something Went Wrong");
    }
  });
};

/**
 * Use Query to Fetch Spell By Index
 * @param {string} id
 * @export useGetSpellByIndex
 * @returns UseQueryResult
 */
export const useGetSpellByIndex = (id: string) => {
  return useQuery([api.spells.fetch, id], getSpellByIndex(id), {
    enabled: !!id,
    select: ({ data }) => data,
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toastFail(error?.response?.data?.message || "Something Went Wrong");
    }
  });
};
