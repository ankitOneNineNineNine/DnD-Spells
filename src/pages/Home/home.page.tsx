import styles from "./home.module.css";
import { useGetAllSpells } from "../../services/spells";
import { Chip } from "../../components/Chip/chip.component";
import { useContext, useEffect, useRef, useState } from "react";
import { FavoriteContext } from "../../infrastructure/Provider/Context";
import { NavLink } from "react-router-dom";
import _ from "lodash";
import { Favorites } from "../Favorites/favorites.pages";
import { Search } from "../../components/Search/search.component";

/**
 * Home Page
 * @export Home
 * @returns JSX.Element
 */
const Home = () => {
  const { data: spellResults, isLoading, isError } = useGetAllSpells();
  const favoriteContext = useContext(FavoriteContext);
  const homePageRef = useRef<HTMLDivElement | null>(null);
  const [showFavoriteModel, setShowFavoriteModel] = useState(false);
  const [searchQuery, setSearchQuery] = useState("");

  /**
   * debounced search (1s after user stops typing)
   */
  const debouncedSearch = useRef(
    _.debounce((searchTerm) => {
      setSearchQuery(searchTerm);
    }, 1000)
  ).current;

  useEffect(() => {
    return () => {
      debouncedSearch.cancel();
    };
  }, [debouncedSearch]);

  /**
   * Close Modal
   */
  const closeModal = () => {
    setShowFavoriteModel(false);
  };

  /**
   * Open Modal
   */
  const openModal = () => {
    setShowFavoriteModel(true);
  };

  /**
   * Handler to Close Favorites Drawer
   * @param {MouseEvent} event
   */
  const handleCloseDrawer = (event: MouseEvent) => {
    if (
      event?.target &&
      homePageRef &&
      !homePageRef.current?.contains(event?.target as Node)
    ) {
      setShowFavoriteModel(false);
    }
  };
  useEffect(() => {
    document.addEventListener("mousedown", handleCloseDrawer);
    return () => {
      document.removeEventListener("mousedown", handleCloseDrawer);
    };
  }, []);

  const filteredSpells = spellResults?.results?.filter((spells) =>
    spells.name.toLowerCase().includes(searchQuery.toLowerCase())
  );

  return (
    <div>
      <h1>Dungeons & Dragons Spells</h1>
      <button
        aria-label="favorite-btn"
        onClick={openModal}
        className={styles.favoriteButton}
      >
        Favorites
      </button>
      <Search text={searchQuery} onChangeHandler={setSearchQuery} />
      {isError ? (
        <p>Something Went Wrong</p>
      ) : isLoading ? (
        <p>Loading ...</p>
      ) : (
        <div data-testid="all-spells" className={styles.spellContainer}>
          {filteredSpells?.map((result) => (
            <NavLink
              data-testid="chip-nav"
              to={`/spells/${result.index}`}
              key={result.index}
            >
              <Chip
                favorite={
                  !!favoriteContext?.favoriteSpells.find(
                    (spell) => spell.index === result.index
                  )
                }
                text={result.name}
                index={result.index}
                onHeartClick={favoriteContext?.onChangeWishList}
              />
            </NavLink>
          ))}
        </div>
      )}

      {showFavoriteModel && (
        <div ref={homePageRef}>
          <Favorites closeModal={closeModal} />
        </div>
      )}
    </div>
  );
};

export default Home;
