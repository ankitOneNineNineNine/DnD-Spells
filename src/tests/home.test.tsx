import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import Home from "../pages/Home/home.page";
import { ProviderMock } from "../mocks/provider.mock";
import { server } from "../mocks/server.mock";
import { rest } from "msw";
import { api } from "../services/api";

describe("home", () => {
  function prepare() {
    const { container } = render(
      <ProviderMock favorites={[]}>
        <Home />
      </ProviderMock>
    );
    return { container };
  }

  it("testing match snapshot", () => {
    const { container } = prepare();
    expect(container).toMatchSnapshot();
  });

  it("open and close favorites component", async () => {
    prepare();
    const favoriteBtn = screen.getByRole("button", { name: /favorite-btn/i });
    fireEvent.click(favoriteBtn);
    expect.assertions(2);
    await waitFor(() => {
      const fav = screen.getByText("Favorite Spells");
      expect(fav).toBeInTheDocument();
    });
    const closeModal = screen.getByTestId("close-modal");
    fireEvent.click(closeModal);

    await waitFor(() => {
      const fav = screen.queryByText("Favorite Spells");
      expect(fav).toBe(null);
    });
  });

  it("mock query response check", async () => {
    prepare();
    expect.assertions(3);
    expect(await screen.findByTestId("all-spells")).toBeInTheDocument();
    await waitFor(() => {
      expect(screen.queryByText("Loading ...")).not.toBeInTheDocument();
    });
    expect(screen.queryByTestId("all-spells")?.childNodes.length).not.toBe(0);
  });

  it("check page change on clicking a chip", async () => {
    prepare();
    const currentLink = window.location.href;
    const linkToChip = await screen.findAllByTestId("chip-nav");
    fireEvent.click(linkToChip[0]);
    expect.assertions(1);
    await waitFor(() => {
      expect(window.location.href).not.toBe(currentLink);
    });
  });
  it("check empty response", async () => {
    server.use(
      rest.get(api.spells.fetch, async (req, res, ctx) => {
        return res(
          ctx.json({
            result: [],
            count: 0,
          })
        );
      })
    );
    prepare();
    await waitFor(() => {
      expect(screen.queryByText("Loading ...")).not.toBeInTheDocument();
    });
    expect(screen.queryByTestId("all-spells")?.childNodes.length).toBe(0);
  });
  it("check error response", async () => {
    server.use(
      rest.get(api.spells.fetch, async (req, res, ctx) => {
        return res(ctx.status(500));
      })
    );
    prepare();
    await waitFor(() => {
      expect(screen.getByText("Something Went Wrong")).toBeInTheDocument();
    });
  });
});
