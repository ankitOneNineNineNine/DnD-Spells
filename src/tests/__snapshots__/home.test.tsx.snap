// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home testing match snapshot 1`] = `
<div>
  <div>
    <h1>
      Dungeons & Dragons Spells
    </h1>
    <button
      aria-label="favorite-btn"
      class="favoriteButton"
    >
      Favorites
    </button>
    <div
      class="search-container"
    >
      <input
        placeholder="Search Spells Here ..."
        type="search"
        value=""
      />
    </div>
    <p>
      Loading ...
    </p>
  </div>
</div>
`;
